universal/include/userver/formats/universal/universal.hpp

```diff
@@ -0,0 +1,289 @@
+#if __cplusplus >= 202002L
+#pragma once
+#include <userver/utils/constexpr_string.hpp>
+#include <userver/formats/parse/to.hpp>
+#include <userver/formats/parse/try_parse.hpp>
+#include <userver/formats/serialize/to.hpp>
+#include <userver/utils/meta.hpp>
+#include <userver/utils/overloaded.hpp>
+#include <variant>
+#include <type_traits>
+#include <boost/pfr/core.hpp>
+#include <boost/pfr/core_name.hpp>
+
+
+USERVER_NAMESPACE_BEGIN
+
+namespace formats::universal {
+
+
+template <typename T, typename Enable = void>
+struct FieldConfig {
+  constexpr inline std::optional<std::string> Check(const T&) const {
+    return std::nullopt;
+  }
+  constexpr auto Write(const T& value, std::string_view field_name, const auto&, auto& builder) const {
+    builder[static_cast<std::string>(field_name)] = value;
+  }
+  template <typename MainClass, auto I, typename Value>
+  constexpr T Read(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    return value[name].template As<T>();
+  }
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<T> TryRead(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    return parse::TryParse(value[name], parse::To<T>{});
+  }
+};
+namespace impl {
+
+template <typename T, std::size_t I>
+using kFieldTypeOnIndex = std::remove_reference_t<decltype(boost::pfr::get<I>(std::declval<T>()))>;
+
+template <typename T>
+consteval std::size_t getFieldIndexByName(std::string_view field_name) {
+  constexpr auto names = boost::pfr::names_as_array<T>();
+  return std::find(names.begin(), names.end(), field_name) - names.begin();
+}
+
+struct Disabled {};
+
+
+template <typename T>
+struct Param {
+  T value;
+  constexpr Param(const T& value) :
+      value(value) {}
+};
+
+template <auto Value>
+struct Wrapper {
+  static constexpr auto kValue = Value;
+};
+template <typename... Args>
+inline constexpr std::optional<std::string> Add(Args&&...) {
+  return std::nullopt;
+}
+
+template <typename Arg, typename... Args>
+inline constexpr std::optional<std::string> Add(Arg&& arg, Args&&... args) {
+  if(arg) {
+    auto toAdd = Add(std::forward<Args>(args)...);
+    if(toAdd) {
+      return *arg + *toAdd;
+    }
+    return *arg;
+  }
+  return Add(std::forward<Arg>(args)...);
```

Comment:
Recursive templates take O(N^2) time to compile, let's rewrite it using fold expressions and comma operator:

```
std::optional<std::string> result;
(..., (args ? ((result ? result.emplace() : void()), result->append(*args)) : void());
return result;
```